File: macos/SeedSeeker/Tests/SeedSeekerKitTests/AutoTrinketTests.swift
import Foundation
import XCTest
@testable import SeedSeekerKit

final class AutoTrinketTests: XCTestCase {
    private func query() throws -> SavedQuery {
        try ResultsExport.decodeQuery(["max_depth": 19, "auto_apply_trinket": true,
            "requirements": [["item": "runic_blade", "upgrade": 1, "effect": "Grim"]]])
    }

    func testDefaultsAndExplicitNoneSurviveDocuments() throws {
        let saved = try query()
        XCTAssertTrue(SavedQuery(requirements: []).autoApplyTrinket)
        XCTAssertFalse(try ResultsExport.decodeQuery(["requirements": [["item": "runic_blade"]]]).autoApplyTrinket)
        XCTAssertTrue(try DeepLink.decode(DeepLink.encodeLink(for: saved)).autoApplyTrinket)
        let restored = try ResultsExport.decode(ResultsExport.encode(saved,
            seeds: ["SRU-YSU-QHS", "EYY-RUL-LQG"], appVersion: "test", trinkets: ["parchment_scrap", nil]))
        XCTAssertEqual(restored.trinkets, ["parchment_scrap", nil])
        XCTAssertTrue(restored.query.autoApplyTrinket)
    }

    func testNativeRefineStripsAndRestoresNeededTrinkets() async throws {
        let saved = try query(); let query = try saved.searchRequest()
        let engine = ProductionSeedFinderEngine()
        let matches = try await engine.filterRecipes(query, base: query, recipes: [
            SeedResult(seed: "SRU-YSU-QHS", matchedRequirements: 1, selectedTrinket: "parchment_scrap"),
            SeedResult(seed: "EYY-RUL-LQG", matchedRequirements: 1, selectedTrinket: "parchment_scrap"),
        ])
        XCTAssertEqual(matches.map(\.selectedTrinket), ["parchment_scrap", nil])
        let whip = try ItemRequirement(key: 2, item: XCTUnwrap(ItemCatalog.findById("whip")), upgrade: 0, effect: .oneOf(["Venomous"]), kind: .weapon, upgradeMatch: .any)
        let refined = try SearchRequest(requirements: query.requirements + [whip], maximumDepth: 19, autoApplyTrinket: true)
        let restored = try await engine.filterRecipes(refined, base: query, recipes: [matches[1]])
        XCTAssertEqual(restored.map(\.selectedTrinket), ["parchment_scrap"])
    }
}

File: windows/SeedSeeker.Tests/AutoTrinketTests.cs
using Xunit;

namespace SeedSeeker.Tests;

public sealed class AutoTrinketTests
{
    private static QuerySettings Query() => ResultsExport.DecodeQueryDocument("""
        {"max_depth":19,"auto_apply_trinket":true,"requirements":[{"item":"runic_blade","upgrade":1,"effect":"Grim"}]}
        """);

    [Fact]
    public void DefaultsAndRecipesSurviveDocumentsAndClones()
    {
        var query = Query();
        Assert.True(query.Clone().AutoApplyTrinket);
        Assert.All(BuiltInPresets.All, preset => Assert.True(preset.Query.AutoApplyTrinket));
        Assert.False(ResultsExport.DecodeQueryDocument("""{"requirements":[{"item":"runic_blade"}]}""").AutoApplyTrinket);
        var restored = ResultsExport.Decode(ResultsExport.Encode(query, ["SRU-YSU-QHS", "EYY-RUL-LQG"], "test", ["parchment_scrap", null]));
        Assert.True(restored.Query.AutoApplyTrinket);
        Assert.Equal(new string?[] { "parchment_scrap", null }, restored.Trinkets);
    }

    [Fact]
    public void NativeFilterStripsOnlyUnnecessaryRecipesAndRefineRestoresNeededChoice()
    {
        var engine = new NativeEngine(); var query = Query();
        var matches = engine.FilterRecipes(query, query, [new("SRU-YSU-QHS", 1, "parchment_scrap"), new("EYY-RUL-LQG", 2, "parchment_scrap")]);
        Assert.Equal(new string?[] { "parchment_scrap", null }, matches.Select(recipe => recipe.SelectedTrinket));
        foreach (var match in matches)
        {
            var trinket = match.SelectedTrinket ?? "none";
            var marked = NativeEngine.ScoutMatches(match.Seed, 0, query, trinket);
            Assert.Equal(marked.TotalRequirements, marked.MatchedRequirements);
        }
        var refined = query.Clone(); refined.Requirements.Add(new() { Item = ItemCatalog.Find("whip"), Effect = EffectFilter.OneOf(["Venomous"]), Kind = ItemKind.Weapon, UpgradeMatch = UpgradeMatch.Any });
        var restored = Assert.Single(engine.FilterRecipes(refined, query, [matches[1]]));
        Assert.Equal("parchment_scrap", restored.SelectedTrinket);
        Assert.Equal(0, NativeEngine.ScoutMatches(matches[0].Seed, 0, query, "none").MatchedRequirements);
    }
}
